File: app/src/main/java/com/gnoemes/shikimoriapp/presentation/presenter/search/SearchPaginatorImpl.java
package com.gnoemes.shikimoriapp.presentation.presenter.search;

import com.gnoemes.shikimoriapp.domain.search.SearchInteractor;
import com.gnoemes.shikimoriapp.entity.anime.domain.Anime;
import com.gnoemes.shikimoriapp.entity.app.data.AppConfig;
import com.gnoemes.shikimoriapp.entity.search.domain.FilterItem;
import com.gnoemes.shikimoriapp.presentation.presenter.common.ViewController;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;

import io.reactivex.disposables.Disposable;

public class SearchPaginatorImpl implements SearchPaginator {

    private SearchInteractor interactor;
    private ViewController<Anime> viewController;

    private int currentPage;
    private int defaultPage = 1;

    private State currentState = new EMPTY_STATE();
    private List<Anime> currentData = Collections.emptyList();
    private HashMap<String, List<FilterItem>> filters = null;
    private Disposable disposable;

    public SearchPaginatorImpl(SearchInteractor interactor, ViewController<Anime> viewController) {
        this.interactor = interactor;
        this.viewController = viewController;
    }

    private void loadPage(int page) {
        unsubscribe();

        disposable = interactor.loadListWithFilters(filters, page, AppConfig.DEFAULT_LIMIT)
                .subscribe(
                        items -> currentState.newData(items),
                        throwable -> currentState.error(throwable));

    }

    private void unsubscribe() {
        if (disposable != null) {
            if (!disposable.isDisposed()) {
                disposable.dispose();
            }
        }
    }

    @Override
    public void setFilters(HashMap<String, List<FilterItem>> filters) {
        this.filters = filters;
    }

    @Override
    public boolean isFirstPage() {
        return currentPage == defaultPage;
    }


    private void increasePage() {
        currentPage = currentPage + 1;
    }

    @Override
    public void restart() {
        currentState.restart();
    }

    @Override
    public void refresh() {
        currentState.refresh();
    }

    @Override
    public void loadNewPage() {
        currentState.loadNewPage();
    }

    @Override
    public void release() {
        currentState.release();
    }


    private interface State {
        void restart();

        void refresh();

        void loadNewPage();

        void release();

        void newData(List<Anime> list);

        void error(Throwable throwable);
    }

    private abstract class EMPTY implements State {
        @Override
        public void restart() {
        }

        @Override
        public void loadNewPage() {
        }

        @Override
        public void newData(List<Anime> list) {
        }

        @Override
        public void error(Throwable throwable) {
        }

    }

    private class EMPTY_STATE extends EMPTY {

        @Override
        public void refresh() {
            currentState = new EMPTY_PROGRESS_STATE();
            viewController.showEmptyProgress(true);
            loadPage(defaultPage);
        }

        @Override
        public void release() {
            currentState = new RELEASED_STATE();
            unsubscribe();
        }
    }

    private abstract class EMPTY_PROGRESS implements State {


        @Override
        public void loadNewPage() {
        }

        @Override
        public void refresh() {

        }

    }

    private class EMPTY_PROGRESS_STATE extends EMPTY_PROGRESS {

        @Override
        public void restart() {
            loadPage(defaultPage);
        }

        @Override
        public void newData(List<Anime> list) {
            if (!list.isEmpty()) {
                currentState = new DATA_STATE();
                currentData.clear();
                currentData = list;
                currentPage = defaultPage;
                viewController.showList(true, currentData);
                viewController.showEmptyView(false);
                viewController.showEmptyProgress(false);
                viewController.showRefreshProgress(false);
            } else {
                currentState = new EMPTY_DATA_STATE();
                viewController.showEmptyProgress(false);
                viewController.showRefreshProgress(false);
                viewController.showEmptyView(true);
            }
        }

        @Override
        public void error(Throwable throwable) {
            currentState = new EMPTY_ERROR_STATE();
            viewController.showEmptyProgress(false);
            viewController.showEmptyError(true, throwable);
            viewController.showRefreshProgress(false);
        }

        @Override
        public void release() {
            currentState = new RELEASED_STATE();
            unsubscribe();
        }
    }

    private abstract class EMPTY_ERROR implements State {
        @Override
        public void loadNewPage() {

        }

        @Override
        public void newData(List<Anime> list) {

        }

        @Override
        public void error(Throwable throwable) {

        }
    }

    private class EMPTY_ERROR_STATE extends EMPTY_ERROR {

        @Override
        public void restart() {
            currentState = new EMPTY_PROGRESS_STATE();
            viewController.showEmptyError(false, null);
            viewController.showEmptyProgress(true);
            loadPage(defaultPage);
        }

        @Override
        public void refresh() {
            currentState = new EMPTY_PROGRESS_STATE();
            viewController.showEmptyError(false, null);
            viewController.showEmptyProgress(true);
            loadPage(defaultPage);
        }

        @Override
        public void release() {
            currentState = new RELEASED_STATE();
            unsubscribe();
        }
    }

    private abstract class EMPTY_DATA implements State {
        @Override
        public void loadNewPage() {

        }

        @Override
        public void newData(List<Anime> list) {

        }

        @Override
        public void error(Throwable throwable) {

        }
    }

    private class EMPTY_DATA_STATE extends EMPTY_DATA {

        @Override
        public void restart() {
            currentState = new EMPTY_PROGRESS_STATE();
            viewController.showEmptyView(false);
            viewController.showRefreshProgress(false);
            viewController.showEmptyProgress(true);
            loadPage(defaultPage);

        }

        @Override
        public void refresh() {
            currentState = new EMPTY_PROGRESS_STATE();
            viewController.showEmptyView(false);
            viewController.showEmptyProgress(true);
            loadPage(defaultPage);

        }

        @Override
        public void release() {
            currentState = new RELEASED_STATE();
            unsubscribe();
        }
    }

    private abstract class DATA implements State {
        @Override
        public void newData(List<Anime> list) {

        }
    }

    private class DATA_STATE extends DATA {

        @Override
        public void restart() {
            currentState = new EMPTY_PROGRESS_STATE();
            viewController.showList(false, null);
            viewController.showRefreshProgress(false);
            viewController.showEmptyProgress(true);
            loadPage(defaultPage);
        }

        @Override
        public void refresh() {
            currentState = new REFRESH_STATE();
            viewController.showRefreshProgress(true);
            loadPage(defaultPage);
        }

        @Override
        public void loadNewPage() {
            currentState = new PAGE_PROGRESS_STATE();
            viewController.showPageProgress(true);
            increasePage();
            loadPage(currentPage);

        }

        @Override
        public void release() {
            currentState = new RELEASED_STATE();
            unsubscribe();
        }

        @Override
        public void error(Throwable throwable) {
            viewController.showError(throwable);
        }
    }


    private abstract class REFRESH implements State {
        @Override
        public void refresh() {

        }

        @Override
        public void loadNewPage() {

        }
    }

    private class REFRESH_STATE extends REFRESH {

        @Override
        public void restart() {
            currentState = new EMPTY_PROGRESS_STATE();
            viewController.showList(false, null);
            viewController.showRefreshProgress(false);
            viewController.showEmptyProgress(true);
            loadPage(defaultPage);
        }

        @Override
        public void newData(List<Anime> list) {
            if (!list.isEmpty()) {
                currentState = new DATA_STATE();
                currentData.clear();
                currentData = list;
                currentPage = defaultPage;
                viewController.showRefreshProgress(false);
                viewController.showEmptyError(false, null);
                viewController.showList(true, currentData);
                viewController.showEmptyView(false);
            } else {
                currentState = new EMPTY_DATA_STATE();
                currentData.clear();
                viewController.showList(false, null);
                viewController.showRefreshProgress(false);
                viewController.showEmptyView(true);
            }
        }

        @Override
        public void error(Throwable throwable) {
            currentState = new DATA_STATE();
            viewController.showRefreshProgress(false);
            viewController.showError(throwable);
        }

        @Override
        public void release() {
            currentState = new RELEASED_STATE();
            unsubscribe();
        }
    }

    private abstract class PAGE_PROGRESS implements State {

        @Override
        public void loadNewPage() {

        }
    }

    private class PAGE_PROGRESS_STATE extends PAGE_PROGRESS {

        @Override
        public void restart() {
            currentState = new EMPTY_PROGRESS_STATE();
            viewController.showList(false, null);
            viewController.showPageProgress(false);
            viewController.showEmptyProgress(true);
            loadPage(defaultPage);
        }

        @Override
        public void newData(List<Anime> list) {
            if (!list.isEmpty()) {
                currentState = new DATA_STATE();
                currentData = list;
                increasePage();
                viewController.showPageProgress(false);
                viewController.showList(true, currentData);
                viewController.showEmptyView(false);
            } else {
                currentState = new ALL_DATA_STATE();
                viewController.showPageProgress(false);
            }

        }

        @Override
        public void refresh() {
            currentState = new REFRESH_STATE();
            viewController.showPageProgress(false);
            viewController.showRefreshProgress(true);
            loadPage(defaultPage);
        }


        @Override
        public void error(Throwable throwable) {
            currentState = new DATA_STATE();
            viewController.showPageProgress(false);
            viewController.showError(throwable);
        }

        @Override
        public void release() {
            currentState = new RELEASED_STATE();
            unsubscribe();
        }
    }


    private abstract class ALL_DATA implements State {

        @Override
        public void newData(List<Anime> list) {

        }

        @Override
        public void error(Throwable throwable) {

        }

        @Override
        public void loadNewPage() {

        }
    }

    private class ALL_DATA_STATE extends ALL_DATA {

        @Override
        public void restart() {
            currentState = new EMPTY_PROGRESS_STATE();
            viewController.showList(false, null);
            viewController.showEmptyProgress(true);
            loadPage(defaultPage);

        }

        @Override
        public void refresh() {
            currentState = new REFRESH_STATE();
            viewController.showRefreshProgress(true);
            loadPage(defaultPage);

        }

        @Override
        public void release() {
            currentState = new RELEASED_STATE();
            unsubscribe();
        }
    }

    private abstract class RELEASED implements State {
        @Override
        public void restart() {

        }

        @Override
        public void refresh() {

        }

        @Override
        public void loadNewPage() {

        }

        @Override
        public void release() {

        }

        @Override
        public void newData(List<Anime> list) {

        }

        @Override
        public void error(Throwable throwable) {

        }
    }

    private class RELEASED_STATE extends RELEASED {

    }
}

File: app/src/main/java/com/gnoemes/shikimoriapp/presentation/presenter/search/SearchPaginatorImpl.kt
package com.gnoemes.shikimoriapp.presentation.presenter.search

import com.gnoemes.shikimoriapp.domain.search.SearchInteractor
import com.gnoemes.shikimoriapp.entity.app.domain.LinkedContent
import com.gnoemes.shikimoriapp.entity.app.domain.Type
import com.gnoemes.shikimoriapp.entity.search.domain.FilterItem
import com.gnoemes.shikimoriapp.presentation.presenter.common.ViewController
import com.gnoemes.shikimoriapp.utils.ifNotNull
import io.reactivex.disposables.Disposable

class SearchPaginatorImpl(private val interactor: SearchInteractor,
                          private val viewController: ViewController<LinkedContent>
) : SearchPaginator {
    private var defaultPage = 1
    private var currentPage = 1

    private var currentState: State<LinkedContent> = EMPTY()
    private var currentData: MutableList<LinkedContent> = mutableListOf()
    private var currentType: Type = Type.ANIME
    var currentFilters: HashMap<String, MutableList<FilterItem>>? = null
    private var disposable: Disposable? = null

    interface State<T> {
        fun restart() = Unit
        fun refresh() = Unit
        fun loadNewPage() = Unit
        fun release() = Unit
        fun newData(list: List<T>) = Unit
        fun error(throwable: Throwable) = Unit
    }

    private fun loadPage(page: Int) {
        when (currentType) {
            Type.ANIME -> loadAnime(page)
            Type.MANGA -> loadManga(page)
            Type.RANOBE -> loadRanobe(page)
        }
    }

    private fun loadRanobe(page: Int) {
        disposable = interactor.loadRanobeListWithFilters(currentFilters, page)
                .subscribe(currentState::newData, currentState::error)
    }

    private fun loadManga(page: Int) {
        disposable = interactor.loadMangaListWithFilters(currentFilters, page)
                .subscribe(currentState::newData, currentState::error)
    }

    private fun loadAnime(page: Int) {
        disposable = interactor.loadAnimeListWithFilters(currentFilters, page)
                .subscribe(currentState::newData, currentState::error)
    }

    private fun unsubscribe() {
        disposable.ifNotNull {
            if (!it.isDisposed) {
                it.dispose()
            }
        }
    }

    override fun setFilters(filters: HashMap<String, MutableList<FilterItem>>) {
        this.currentFilters = filters
    }

    override fun setType(type: Type) {
        currentType = type
    }

    override fun isFirstPage(): Boolean = currentPage == defaultPage
    override fun loadNewPage() = currentState.loadNewPage()
    override fun restart() = currentState.restart()
    override fun refresh() = currentState.refresh()
    override fun release() = currentState.release()
    private fun increasePage() {
        currentPage += 1
    }

    private inner class EMPTY : State<LinkedContent> {
        override fun refresh() {
            currentState = EMPTY_PROGRESS()
            viewController.showEmptyProgress(true)
            loadPage(defaultPage)
        }

        override fun release() {
            currentState = RELEASED()
            unsubscribe()
        }
    }

    private inner class EMPTY_PROGRESS : State<LinkedContent> {
        override fun restart() {
            loadPage(defaultPage)
        }

        override fun newData(list: List<LinkedContent>) {
            if (!list.isEmpty()) {
                currentState = DATA()
                currentData.clear()
                currentData = list.toMutableList()
                currentPage = defaultPage
                viewController.showList(true, currentData)
                viewController.showEmptyView(false)
                viewController.showEmptyProgress(false)
                viewController.showRefreshProgress(false)
            } else {
                currentState = EMPTY_DATA()
                viewController.showEmptyProgress(false)
                viewController.showRefreshProgress(false)
                viewController.showEmptyView(true)
            }
        }

        override fun error(throwable: Throwable) {
            currentState = EMPTY_ERROR()
            viewController.showEmptyProgress(false)
            viewController.showEmptyError(true, throwable)
            viewController.showRefreshProgress(false)
        }

        override fun release() {
            currentState = RELEASED()
            unsubscribe()
        }
    }

    private inner class EMPTY_ERROR : State<LinkedContent> {
        override fun restart() {
            currentState = EMPTY_PROGRESS()
            viewController.showEmptyError(false, null)
            viewController.showEmptyProgress(true)
            loadPage(defaultPage)
        }

        override fun refresh() {
            currentState = EMPTY_PROGRESS()
            viewController.showEmptyError(false, null)
            viewController.showEmptyProgress(true)
            loadPage(defaultPage)
        }

        override fun release() {
            currentState = RELEASED()
            unsubscribe()
        }
    }

    private inner class EMPTY_DATA : State<LinkedContent> {
        override fun restart() {
            currentState = EMPTY_PROGRESS()
            viewController.showEmptyView(false)
            viewController.showRefreshProgress(false)
            viewController.showEmptyProgress(true)
            loadPage(defaultPage)

        }

        override fun refresh() {
            currentState = EMPTY_PROGRESS()
            viewController.showEmptyView(false)
            viewController.showEmptyProgress(true)
            loadPage(defaultPage)

        }

        override fun release() {
            currentState = RELEASED()
            unsubscribe()
        }
    }

    private inner class DATA : State<LinkedContent> {
        override fun restart() {
            currentState = EMPTY_PROGRESS()
            viewController.showList(false, null)
            viewController.showRefreshProgress(false)
            viewController.showEmptyProgress(true)
            loadPage(defaultPage)
        }

        override fun refresh() {
            currentState = REFRESH()
            viewController.showRefreshProgress(true)
            loadPage(defaultPage)
        }

        override fun loadNewPage() {
            currentState = PAGE_PROGRESS()
            viewController.showPageProgress(true)
            increasePage()
            loadPage(currentPage)

        }

        override fun release() {
            currentState = RELEASED()
            unsubscribe()
        }

        override fun error(throwable: Throwable) {
            viewController.showError(throwable)
        }
    }

    private inner class REFRESH : State<LinkedContent> {
        override fun restart() {
            currentState = EMPTY_PROGRESS()
            viewController.showList(false, null)
            viewController.showRefreshProgress(false)
            viewController.showEmptyProgress(true)
            loadPage(defaultPage)
        }

        override fun newData(list: List<LinkedContent>) {
            if (!list.isEmpty()) {
                currentState = DATA()
                currentData.clear()
                currentData = list.toMutableList()
                currentPage = defaultPage
                viewController.showRefreshProgress(false)
                viewController.showEmptyError(false, null)
                viewController.showList(true, currentData)
                viewController.showEmptyView(false)
            } else {
                currentState = EMPTY_DATA()
                currentData.clear()
                viewController.showList(false, null)
                viewController.showRefreshProgress(false)
                viewController.showEmptyView(true)
            }
        }

        override fun error(throwable: Throwable) {
            currentState = DATA()
            viewController.showRefreshProgress(false)
            viewController.showError(throwable)
        }

        override fun release() {
            currentState = RELEASED()
            unsubscribe()
        }
    }

    private inner class PAGE_PROGRESS : State<LinkedContent> {
        override fun restart() {
            currentState = EMPTY_PROGRESS()
            viewController.showList(false, null)
            viewController.showPageProgress(false)
            viewController.showEmptyProgress(true)
            loadPage(defaultPage)
        }

        override fun newData(list: List<LinkedContent>) {
            if (!list.isEmpty()) {
                currentState = DATA()
                currentData = list.toMutableList()
                viewController.showPageProgress(false)
                viewController.showList(true, currentData)
                viewController.showEmptyView(false)
            } else {
                currentState = ALL_DATA()
                viewController.showPageProgress(false)
            }

        }

        override fun refresh() {
            currentState = REFRESH()
            viewController.showPageProgress(false)
            viewController.showRefreshProgress(true)
            loadPage(defaultPage)
        }


        override fun error(throwable: Throwable) {
            currentState = DATA()
            viewController.showPageProgress(false)
            viewController.showError(throwable)
        }

        override fun release() {
            currentState = RELEASED()
            unsubscribe()
        }
    }

    private inner class ALL_DATA : State<LinkedContent> {
        override fun restart() {
            currentState = EMPTY_PROGRESS()
            viewController.showList(false, null)
            viewController.showEmptyProgress(true)
            loadPage(defaultPage)

        }

        override fun refresh() {
            currentState = REFRESH()
            viewController.showRefreshProgress(true)
            loadPage(defaultPage)

        }

        override fun release() {
            currentState = RELEASED()
            unsubscribe()
        }
    }

    private inner class RELEASED : State<LinkedContent>

}